[iOS] Add a (off-by-default) setting to make bidi text selections appear visually contiguous

whsieh opened the pull request:
#### fc02c970c4f7dd90289db2a822a60e38581556aa
<pre>
[iOS] Add a (off-by-default) setting to make bidi text selections appear visually contiguous
<a href="https://bugs.webkit.org/show_bug.cgi?id=281100">https://bugs.webkit.org/show_bug.cgi?id=281100</a>

Reviewed by Richard Robinson.

Adjust some iOS text selection codepaths to allow for visually-contiguous selection views (that
appear logically discontiguous); guard this new behavior behind a feature flag, which is off by
default. This is work towards implementing a more intuitive selection behaviors for bidi text, by
ensuring visual stability while the user is interacting with selection handles.

In a subsequent patch, we&apos;ll extend the selection to ensure both logical and visual contiguity after
the user is done selecting text.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-expected.txt: Added.
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection.html: Added.

Add a basic layout test to ensure that selection UI remains stable while the user is in the process
of selecting text.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add the new internal feature flag.

* Source/WebCore/platform/ios/SelectionGeometry.cpp:
(WebCore::SelectionGeometry::SelectionGeometry):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/ios/SelectionGeometry.h:
(WebCore::SelectionGeometry::mayAppearLogicallyDiscontiguous const):
(WebCore::SelectionGeometry::setMayAppearLogicallyDiscontiguous):

Add a new boolean flag to indicate that the `SelectionGeometry` may have a logically discontiguous
appearance.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::areOnSameLine):
(WebCore::makeBidiSelectionVisuallyContiguousIfNeeded):

Implement the main heuristic to ensure visual contiguity here. At a high level, this post-processes
coalesced selection geometries such that:

-   Interior selection geometries (neither the start nor end) are unaltered.
-   Geometries that contain the start are all coalesced into a single selection that extends from
    the selection start caret rect, to the end of the line containing the start position.
-   Likewise, geometries that contain the end are coalesced into a selection that extends from the
    start of the line containing the end position, to the end caret rect.

If all of the selected text is left-to-right, we simply skip the steps above. For now, we also only
support horizontal bidi text (i.e. bail from this codepath if there are any vertical selections).

(WebCore::RenderObject::collectSelectionGeometriesInternal):
(WebCore::RenderObject::collectSelectionGeometries):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm:
(-[WKTextSelectionRect writingDirection]):

Canonical link: <a href="https://commits.webkit.org/284907@main">https://commits.webkit.org/284907@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea7e39db714d54f5771088377e22733a52cf8487

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50267 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74965 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22058 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72973 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58065 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21878 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14554 "layout-tests (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45700 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61095 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36537 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42353 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18529 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20406 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/63987 "Found 1 new JSC binary failure: testapi (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64289 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18890 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76682 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70106 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18094 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15131 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61153 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63766 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15693 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11842 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5490 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91887 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46068 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/839 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20027 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47140 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48421 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46882 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2520,12 +2520,137 @@ Vector<SelectionGeometry> RenderObject::collectSelectionGeometriesWithoutUnionIn
+
+    std::optional<SelectionGeometry> startGeometry;
+    std::optional<SelectionGeometry> endGeometry;
+    for (size_t i = 0; i < geometries.size(); ++i) {
```
can this use a `for const auto& ... : ...` loop instead?

whsieh replied:
Oh, good catch — yes, this should be an iterator for loop since I don't actually need the index (anymore).

Thanks for the review!

rniwa commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -7919,6 +7919,19 @@ VisualViewportEnabled:
     WebCore:
       default: true
 
+VisuallyContiguousBidiTextSelectionEnabled:
```
Can we add conditional: PLATFORM(ISO_FAMILY) given it looks like this option only works in IOS_FAMILY?

whsieh replied:
Sounds good, will add `conditional: PLATFORM(IOS_FAMILY)`.

rniwa commented on LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection.html:
```diff
@@ -0,0 +1,54 @@
+</script>
+</head>
+<body>
+    <p>Arabic, <span class="stop">مثل</span> هذا النص, is right to <span class="start">left</span>.</p>
```
Can we add a ref test which asserts that the selections are visually contiguous?
Also, what happens to selection set by scripts? We don't show them visually contiguous?

whsieh replied:
> Can we add a ref test which asserts that the selections are visually contiguous?

Hm... a ref test is tricky since selections don't render as part of web content (and thus, don't appear in the reference snapshots). However, I think I can achieve a similar result in one of two ways:
1. By adding some JavaScript helper to check for visual discontinuity, or:
2. By adding some JavaScript helper to append absolutely-positioned DOM elements to the page over the selection rects (and then use ref-testing).

> Also, what happens to selection set by scripts? We don't show them visually contiguous?

Right, we'd want to preserve existing behavior and just fall back to a visually discontiguous (but logically contiguous) selection. I haven't gotten to implementing that behavior yet — it'll be in a later patch.

whsieh replied:
For now, I think I'll implement the helper function in (1) inside `ui-helper.js`, and use it here so that the expectation is something like:

```
PASS selectionIsVisuallyContiguous is true
```

whsieh committed:
[iOS] Add a (off-by-default) setting to make bidi text selections appear visually contiguous
https://bugs.webkit.org/show_bug.cgi?id=281100

Reviewed by Richard Robinson.

Adjust some iOS text selection codepaths to allow for visually-contiguous selection views (that
appear logically discontiguous); guard this new behavior behind a feature flag, which is off by
default. This is work towards implementing a more intuitive selection behaviors for bidi text, by
ensuring visual stability while the user is interacting with selection handles.

In a subsequent patch, we'll extend the selection to ensure both logical and visual contiguity after
the user is done selecting text.

* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection-expected.txt: Added.
* LayoutTests/editing/selection/ios/bidi-visually-contiguous-selection.html: Added.

Add a basic layout test to ensure that selection UI remains stable while the user is in the process
of selecting text.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add the new internal feature flag.

* Source/WebCore/platform/ios/SelectionGeometry.cpp:
(WebCore::SelectionGeometry::SelectionGeometry):
(WebCore::operator<<):
* Source/WebCore/platform/ios/SelectionGeometry.h:
(WebCore::SelectionGeometry::mayAppearLogicallyDiscontiguous const):
(WebCore::SelectionGeometry::setMayAppearLogicallyDiscontiguous):

Add a new boolean flag to indicate that the `SelectionGeometry` may have a logically discontiguous
appearance.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::areOnSameLine):
(WebCore::makeBidiSelectionVisuallyContiguousIfNeeded):

Implement the main heuristic to ensure visual contiguity here. At a high level, this post-processes
coalesced selection geometries such that:

-   Interior selection geometries (neither the start nor end) are unaltered.
-   Geometries that contain the start are all coalesced into a single selection that extends from
    the selection start caret rect, to the end of the line containing the start position.
-   Likewise, geometries that contain the end are coalesced into a selection that extends from the
    start of the line containing the end position, to the end caret rect.

If all of the selected text is left-to-right, we simply skip the steps above. For now, we also only
support horizontal bidi text (i.e. bail from this codepath if there are any vertical selections).

(WebCore::RenderObject::collectSelectionGeometriesInternal):
(WebCore::RenderObject::collectSelectionGeometries):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WKTextSelectionRect.mm:
(-[WKTextSelectionRect writingDirection]):

Canonical link: https://commits.webkit.org/284907@main
